pandas/core/dtypes/cast.py

```diff
@@ -2096,6 +2096,9 @@ def maybe_cast_to_integer_array(
         )
         return casted
 
+    if all(np.dtype(i) is dtype for i in casted):
```

Comment:
this is going to be super-slow

Reply:
can maybe update the comment on L2102 if we've found cases that get here

Reply:
I don't recall any cases that reach the ValueError that were there. There was one case which the added code solved.

Reply:
I do not have a faster method to check whether all the cast for all the elements is performed correctly. 

Checking the dtypes is one way but individually we cannot check all elements.

This is executed last, so in many cases we will not reach here. So a bit slow can do.

Reply:
yeah we need to be really sure about this check

Reply:
Changed the check as requested.